docker_run.sh

```diff
@@ -53,20 +53,48 @@ else
     cd src/
     ln -s ../build CMakeFiles
     cd ..
+    
+    BUILD_START=$(date)
+    echo $BUILD_START
+    
 # run cmake separately so config.h will be seen by the md5 script
     cmake -DCMAKE_BUILD_TYPE=Release -DENABLE_GAME=true -DENABLE_SERVER=true -DENABLE_CAMPAIGN_SERVER=true -DENABLE_TESTS=true -DENABLE_NLS=false -DEXTRA_FLAGS_CONFIG="-pipe" -DEXTRA_FLAGS_RELEASE="$EXTRA_FLAGS_RELEASE" -DENABLE_STRICT_COMPILATION="$STRICT"
 # run manual md5 file tracking/mtime modification script for cmake
     python3 cmake_mtime_crc.py
+    PY_RES=$?
+    if [ "$PY_RES" != "0" ]; then
+      rm build/cmake_mtime_crc.txt
+      echo "Failed to update cmake file mtimes!"
+      exit 1
+    fi
     
-    make VERBOSE=1 -j2
+# filter out spammy "is newer than dependee"/"is newer than depends" output that can occasionally fail the build because of its sheer volume
+    make VERBOSE=1 -j2 | grep -v "is newer than"
     
   else
     scons wesnoth wesnothd campaignd boost_unit_tests build=release ctool=$CC cxxtool=$CXX --debug=time extra_flags_config="-pipe" extra_flags_release="$EXTRA_FLAGS_RELEASE" strict="$STRICT" cxx_std=$CXXSTD nls=false jobs=2
   fi
   
-# check if the build was successful
+# check if the build failed
   BUILD_RET=$?
   if [ $BUILD_RET != 0 ]; then
+# if the failed build was using cmake, clear out all the files that were built and remove their stored CRC, to make sure they're all treated as brand new and recompiled on subsequent runs
+# otherwise, it is possible for there to be source files that didn't change between runs, but still need to be recompiled, and they would be missed which would corrupt the cache
+# this is especially possible when there are unrelated commits that don't involve C++ changes made between the original bad commit and the commit fixing the issue
+    if [ "$TOOL" = "cmake" ]; then
+      declare -a FILES=($(find build/ -type f -newermt "$BUILD_START" -name '*.o'))
+      
+      echo "Files being cleaned up after failed build:"
+      for (( i = 0; i < ${#FILES[@]}; i++ )); do
+          OBJECT_FILE=${FILES[$i]}
+          ls -al $OBJECT_FILE
+          rm $OBJECT_FILE
+          
+          SOURCE_FILE='src/'$(echo $OBJECT_FILE | cut -d/ -f3- | cut -f1 -d.)
+          sed -i 's#^'$SOURCE_FILE'.cpp:[0-9]*\|^'$SOURCE_FILE'.hpp:[0-9]*\|^'$SOURCE_FILE'.c:[0-9]*\|^'$SOURCE_FILE'.h:[0-9]*\|^'$SOURCE_FILE'.tpp:[0-9]*##g' build/cmake_mtime_crc.txt
```

Comment:
This looks fragile. It would probably be safer to just trigger a full rebuild next time, like you mentioned on Discord.

Reply:
I wouldn't say it's overly fragile - it just trims off the `build/<something>.dir/, and then removes the file extension - but yeah, a full rebuild would always be safer.